.github/PULL_REQUEST_TEMPLATE.md

```diff
@@ -0,0 +1,32 @@
+<!--
+PLEASE READ THIS MESSAGE.
+
+HOW TO WRITE A GOOD PULL REQUEST?
+
+- Make it small.
+- Do only one thing.
+- Avoid re-formatting.
+- Make sure the code builds.
+- Add tests
+- Write useful descriptions and titles.
+- Address review comments in terms of additional commits.
+- Do not amend/squash existing ones unless the PR is trivial.
```

Comment:
Now that I think about it, I'd rather we squash commits before merging. What do you think of:

`- Once approved, please squash your commits to avoid polluting the commit history`

Reply:
Hmm why squashing?
Personally i'm more fan of rebasing (putting on top of current master, where you can choose if you want to squash some commits or even drop them) instead of squashing (all commits in one single commit)

Reply:
I've updated it to your comment.

I've just thought as per these rules you...
- Should have just one item in your pr
- Can have multiple commits to present the feature/bugfix/etc
- Besides above commits, have review commits which can be squashed together after all review messages have been done (creating a new squashed commit containing the git messages from these changes).
- Run `git rebase -i origin/master` to make sure there's unessairy stuff in.
- Click rebase and merge button in github

Reply:
Apologies, I didn't see your reply. I favor squashing as it results in a clean commit history: there's a one-to-one mapping between each commit and each work item. That being said, the project is mature to have strict rules. So let's stick to whatever works best for each change as they come up?